primitives/core/src/traits.rs

```diff
@@ -52,7 +55,158 @@ pub enum Error {
+		&self,
+		id: KeyTypeId,
+		keys: Vec<CryptoTypePublicPair>
+	) -> Result<Vec<CryptoTypePublicPair>, Error>;
+	/// List all supported keys
+	///
+	/// Returns a set of public keys the signer supports.
+	async fn keys(&self, id: KeyTypeId) -> Result<Vec<CryptoTypePublicPair>, Error>;
+
+	/// Checks if the private keys for the given public key and key type combinations exist.
+	///
+	/// Returns `true` iff all private keys could be found.
+	async fn has_keys(&self, public_keys: &[(Vec<u8>, KeyTypeId)]) -> bool;
+
+	/// Sign with key
+	///
+	/// Signs a message with the private key that matches
+	/// the public key passed.
+	///
+	/// Returns the SCALE encoded signature if key is found & supported,
+	/// an error otherwise.
+	async fn sign_with(
+		&self,
+		id: KeyTypeId,
+		key: &CryptoTypePublicPair,
+		msg: &[u8],
+	) -> Result<Vec<u8>, Error>;
+
+	/// Sign with any key
+	///
+	/// Given a list of public keys, find the first supported key and
+	/// sign the provided message with that key.
+	///
+	/// Returns a tuple of the used key and the SCALE encoded signature.
+	async fn sign_with_any(
+		&self,
+		id: KeyTypeId,
+		keys: Vec<CryptoTypePublicPair>,
+		msg: &[u8]
+	) -> Result<(CryptoTypePublicPair, Vec<u8>), Error> {
+		if keys.len() == 1 {
+			return self.sign_with(id, &keys[0], msg).await.map(|s| (keys[0].clone(), s));
+		} else {
+			for k in self.supported_keys(id, keys).await? {
+				if let Ok(sign) = self.sign_with(id, &k, msg).await {
+					return Ok((k, sign));
+				}
+			}
+		}
+		Err(Error::KeyNotSupported(id))
+	}
+
+	/// Sign with all keys
+	///
+	/// Provided a list of public keys, sign a message with
+	/// each key given that the key is supported.
+	///
+	/// Returns a list of `Result`s each representing the SCALE encoded
+	/// signature of each key or a Error for non-supported keys.
+	async fn sign_with_all(
+		&self,
+		id: KeyTypeId,
+		keys: Vec<CryptoTypePublicPair>,
+		msg: &[u8],
+	) -> Result<Vec<Result<Vec<u8>, Error>>, ()> {
+		let futs = keys.iter()
+			.map(|k| self.sign_with(id, k, msg));
+
+		Ok(join_all(futs).await)
+	}
+
+	/// Generate VRF signature for given transcript data.
+	///
+	/// Receives KeyTypeId and Public key to be able to map
+	/// them to a private key that exists in the keystore which
+	/// is, in turn, used for signing the provided transcript.
+	///
+	/// Returns a result containing the signature data.
+	/// Namely, VRFOutput and VRFProof which are returned
+	/// inside the `VRFSignature` container struct.
+	///
+	/// This function will return an error in the cases where
+	/// the public key and key type provided do not match a private
+	/// key in the keystore. Or, in the context of remote signing
+	/// an error could be a network one.
+	async fn sr25519_vrf_sign(
+		&self,
+		key_type: KeyTypeId,
+		public: &sr25519::Public,
+		transcript_data: VRFTranscriptData,
+	) -> Result<VRFSignature, Error>;
+}
+
+#[allow(missing_docs)]
+pub trait SyncCryptoStore: Send + Sync {
```

Comment:
I saw a neat trick the other day to prevent others from implementing this trait. Maybe we want to employ that here (but we sure want the doc comment!).

```suggestion
mod private {
    // internal trait that only this crate can implement to
    // restrict, who can implement `SyncCryptoStore`.
    trait Sealed {}
}

impl private::Sealed for CryptoStore {};

/// [deprecated] Sync version of the CryptoStore 
///
/// Some parts of Substrate still rely on a sync version of the `CryptoStore`.To make the
/// transition easier this auto trait wraps any async `CryptoStore` and exposes a `sync`
/// interface using `block_on`. Usage of this is deprecated and it will be removed as soon
/// as the internal usage has transitioned successfully. Implementing this `trait` is also made
/// impossible. If you are starting out building something new **do not use this**,
/// but [`CryptoStore`].
///
pub trait SyncCryptoStore: Send + Sync + private::Sealed {
```

Reply:
I changed you version to:
```
mod private {
    // internal trait that only this crate can implement to
    // restrict, who can implement `SyncCryptoStore`.
    pub(crate) trait Sealed {}
}

impl private::Sealed for Arc<dyn CryptoStore> {}
/// [deprecated] Sync version of the CryptoStore
///
/// Some parts of Substrate still rely on a sync version of the `CryptoStore`.To make the
/// transition easier this auto trait wraps any async `CryptoStore` and exposes a `sync`
/// interface using `block_on`. Usage of this is deprecated and it will be removed as soon
/// as the internal usage has transitioned successfully. Implementing this `trait` is also made
/// impossible. If you are starting out building something new **do not use this**,
/// but [`CryptoStore`].
///
pub trait SyncCryptoStore: Send + Sync + private::Sealed {	/// Returns all sr25519 public keys for the given key type.
```

Which marks `Sealed` as `pub(crate)` so that anything outside `private` can implement that. Also switched `impl Sealed for Arc<dyn CryptoStore>` according to recent changes on the code. However, this complains that the private type is leaked through a public interface:
```
error[E0445]: crate-visible trait `Sealed` in public interface
   --> primitives/core/src/traits.rs:224:1
    |
212 |       pub(crate) trait Sealed {}
    |       ---------- `Sealed` declared as crate-visible
...
224 | / pub trait SyncCryptoStore: Send + Sync + private::Sealed {    /// Returns all sr25519 public keys for the given key type.
225 | |     fn sr25519_public_keys(&self, id: KeyTypeId) -> Vec<sr25519::Public>;
226 | |     /// Generate a new sr25519 key pair for the given key type and an optional seed.
227 | |     ///
...   |
361 | |     ) -> Result<VRFSignature, Error>;
362 | | }
    | |_^ can't leak crate-visible trait
```

Reply:
>  so that anything outside private can implement that

Is that a requirement we need?

@rakanalh the following compiles:

```diff
diff --git a/primitives/core/src/traits.rs b/primitives/core/src/traits.rs
index ed675eb75..5159c4c12 100644
--- a/primitives/core/src/traits.rs
+++ b/primitives/core/src/traits.rs
@@ -206,13 +206,19 @@ pub trait CryptoStore: Send + Sync {
        ) -> Result<VRFSignature, Error>;
 }
 
+mod private {
+       pub trait Sealed {}
+
+       impl Sealed for std::sync::Arc<dyn super::CryptoStore> {}
+}
+
 /// A synchronous keystore trait.
 ///
 /// Note: This trait is temporarily maintained until
 /// substrate's crates that use the keystore are fully transitioned
 /// to async/await. Therefore, this trait should not be relied upon
 /// nor used for any new components.
-pub trait SyncCryptoStore: Send + Sync {
+pub trait SyncCryptoStore: Send + Sync + private::Sealed {
        /// Returns all sr25519 public keys for the given key type.
        fn sr25519_public_keys(&self, id: KeyTypeId) -> Vec<sr25519::Public>;
        /// Generate a new sr25519 key pair for the given key type and an optional seed.
```

See https://rust-lang.github.io/api-guidelines/future-proofing.html#c-sealed for details.

Reply:
With my suggestion below (https://github.com/paritytech/substrate/pull/7000#issuecomment-697201486) where `SyncCryptostore` is an extension trait I don't think there is any need for sealing as the trait signature itself enforces that `SyncCryptostore` is only implemented for things implementing `CryptoStore`.